core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala

```diff
@@ -451,59 +490,106 @@ class GroupMetadataManager(brokerId: Int,
 
           memRecords.batches.asScala.foreach { batch =>
-            for (record <- batch.asScala) {
-              require(record.hasKey, "Group metadata/offset entry key should not be null")
-              GroupMetadataManager.readMessageKey(record.key) match {
-
-                case offsetKey: OffsetKey =>
-                  // load offset
-                  val key = offsetKey.key
-                  if (!record.hasValue) {
-                    loadedOffsets.remove(key)
-                    removedOffsets.add(key)
-                  } else {
-                    val value = GroupMetadataManager.readOffsetMessageValue(record.value)
-                    loadedOffsets.put(key, value)
-                    removedOffsets.remove(key)
-                  }
-
-                case groupMetadataKey: GroupMetadataKey =>
-                  // load group metadata
-                  val groupId = groupMetadataKey.key
-                  val groupMetadata = GroupMetadataManager.readGroupMessageValue(groupId, record.value)
-                  if (groupMetadata != null) {
-                    trace(s"Loaded group metadata for group $groupId with generation ${groupMetadata.generationId}")
-                    removedGroups.remove(groupId)
-                    loadedGroups.put(groupId, groupMetadata)
-                  } else {
-                    loadedGroups.remove(groupId)
-                    removedGroups.add(groupId)
+            val isTxnOffsetCommit = batch.isTransactional
+            if (batch.isControlBatch) {
+              val record = batch.iterator.next()
+              val controlRecord = ControlRecordType.parse(record.key)
+              if (controlRecord == ControlRecordType.COMMIT) {
+                val pendingOffsetsFromProducer = pendingOffsets.getOrElse(batch.producerId, null)
+                if (pendingOffsetsFromProducer != null) {
+                  pendingOffsetsFromProducer.foreach { case (groupTopicPartition, offsetAndMetadata) =>
+                    loadedOffsets.put(groupTopicPartition, offsetAndMetadata)
                   }
-
-                case unknownKey =>
-                  throw new IllegalStateException(s"Unexpected message key $unknownKey while loading offsets and group metadata")
+                 pendingOffsets.remove(batch.producerId)
+                }
+              } else {
+                pendingOffsets.remove(batch.producerId)
+              }
+            } else {
+              for (record <- batch.asScala) {
+                require(record.hasKey, "Group metadata/offset entry key should not be null")
+                GroupMetadataManager.readMessageKey(record.key) match {
+
+                  case offsetKey: OffsetKey =>
+                    if (isTxnOffsetCommit && !pendingOffsets.contains(batch.producerId))
+                      pendingOffsets.put(batch.producerId, mutable.Map[GroupTopicPartition, OffsetAndMetadata]())
+
+                    // load offset
+                    val key = offsetKey.key
```

Comment:
nit: can we rename this to something more descriptive. Maybe `groupTopicPartition`?